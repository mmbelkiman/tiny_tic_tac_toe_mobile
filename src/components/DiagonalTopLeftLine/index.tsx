import React, {useEffect, useRef} from 'react';
import {Animated} from 'react-native';

interface DiagonalTopLeftLineProps {
  length: number;
  height: number;
  winner: 'X' | 'O' | null;
  visible: boolean;
}

const DiagonalTopLeftLine: React.FC<DiagonalTopLeftLineProps> = ({
  length,
  height,
  winner,
  visible,
}) => {
  if (!visible || winner === null) {
    return null;
  }

  const lineColor =
    winner === 'X' ? 'rgba(216, 14, 13, 0.8)' : 'rgba(0, 0, 0, 0.8)';
  const opacity = useRef(new Animated.Value(0)).current;
  const heightAnimation = useRef(new Animated.Value(0)).current;
  const animation = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    Animated.timing(opacity, {
      toValue: 0.8,
      duration: 90,
      useNativeDriver: false,
    }).start();

    // Play height animation from 0 to height
    Animated.timing(heightAnimation, {
      toValue: height,
      duration: 80,
      useNativeDriver: false,
    }).start();

    // Play shake animation
    Animated.sequence([
      Animated.timing(animation, {
        toValue: 1,
        duration: 200,
        useNativeDriver: false,
      }),
      Animated.timing(animation, {
        toValue: 0,
        duration: 200,
        useNativeDriver: false,
      }),
    ]).start();
  }, [opacity, heightAnimation, animation, height]);

  return (
    <Animated.View
      style={{
        position: 'absolute',
        backgroundColor: lineColor,
        width: length,
        height: heightAnimation,
        transform: [
          {translateX: -length / 2},
          {translateY: -height / 2},
          {rotate: '45deg'},
          {translateX: length / 2},
          {translateY: height / 2},
          {
            translateY: animation.interpolate({
              inputRange: [0, 1],
              outputRange: [0, -5],
            }),
          },
        ],
        opacity,
      }}
    />
  );
};

export default DiagonalTopLeftLine;
