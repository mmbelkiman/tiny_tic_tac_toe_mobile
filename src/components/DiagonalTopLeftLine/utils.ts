import {Animated} from 'react-native';
import {getPlayerColor} from '@/common/utils';
import {PlayerMark} from '@/common/types';
import {getShakeStyle} from './animations';
import styles from './styles';

export const getLineStyle = ({
  winner,
  length,
  height,
  left,
  heightAnimation,
  opacityAnimation,
  shakeAnimation,
}: {
  winner: PlayerMark;
  length: number;
  height: number;
  left: number;
  heightAnimation: Animated.Value;
  opacityAnimation: Animated.Value;
  shakeAnimation: Animated.Value;
}) => [
  styles.line,
  {
    backgroundColor: getPlayerColor(winner),
    width: length,
    left: left,
    height: heightAnimation,
    opacity: opacityAnimation,
  },
  getShakeStyle(shakeAnimation, length, height),
];
