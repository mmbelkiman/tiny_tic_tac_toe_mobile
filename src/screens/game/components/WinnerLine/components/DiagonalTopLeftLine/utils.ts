import {Animated} from 'react-native';
import styles from './styles';

export const getShakeStyle = (
  animation: Animated.Value,
  length: number,
  height: number,
) => ({
  transform: [
    {translateX: -length / 2},
    {translateY: -height / 2},
    {rotate: '45deg'},
    {translateX: length / 2},
    {translateY: height / 2},
    {
      translateY: animation.interpolate({
        inputRange: [0, 1],
        outputRange: [0, -5],
      }),
    },
  ],
});

export const getLineStyle = ({
  color,
  length,
  height,
  left,
  heightAnimation,
  opacityAnimation,
  shakeAnimation,
}: {
  color: string;
  length: number;
  height: number;
  left: number;
  heightAnimation: Animated.Value;
  opacityAnimation: Animated.Value;
  shakeAnimation: Animated.Value;
}) => [
  styles.line,
  {
    backgroundColor: color,
    width: length,
    left: left,
    height: heightAnimation,
    opacity: opacityAnimation,
  },
  getShakeStyle(shakeAnimation, length, height),
];
