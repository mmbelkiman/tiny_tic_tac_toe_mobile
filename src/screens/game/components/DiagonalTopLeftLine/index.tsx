import React, {useEffect, useRef} from 'react';
import {Animated} from 'react-native';
import {
  playOpacityAnimation,
  playShakeAnimation,
  playSizeAnimation,
} from '@/common/animations';
import {getLineStyle} from './utils';
import {PlayerMark} from '@/screens/game/common/types';

interface DiagonalTopLeftLineProps {
  length: number;
  height: number;
  winner: PlayerMark;
}

const DiagonalTopLeftLine: React.FC<DiagonalTopLeftLineProps> = ({
  length,
  height,
  winner,
}) => {
  const opacityAnimation = useRef(new Animated.Value(0)).current;
  const heightAnimation = useRef(new Animated.Value(0)).current;
  const shakeAnimation = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    if (winner !== '') {
      playOpacityAnimation(opacityAnimation);
      playSizeAnimation(heightAnimation, height);
      playShakeAnimation(shakeAnimation);
    }
  }, [height, heightAnimation, opacityAnimation, shakeAnimation, winner]);

  if (winner === '') {
    return null;
  }

  return (
    <Animated.View
      style={getLineStyle({
        winner,
        length,
        height,
        left: height,
        heightAnimation,
        opacityAnimation,
        shakeAnimation,
      })}
    />
  );
};

export default DiagonalTopLeftLine;
