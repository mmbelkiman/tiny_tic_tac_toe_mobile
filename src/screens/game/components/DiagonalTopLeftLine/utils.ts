import {Animated} from 'react-native';
import {getPlayerColor} from '../../../../common/utils';
import {PlayerMark} from '../../../../common/types';
import styles from './styles';

export const getShakeStyle = (
  animation: Animated.Value,
  length: number,
  height: number,
) => ({
  transform: [
    {translateX: -length / 2},
    {translateY: -height / 2},
    {rotate: '45deg'},
    {translateX: length / 2},
    {translateY: height / 2},
    {
      translateY: animation.interpolate({
        inputRange: [0, 1],
        outputRange: [0, -5],
      }),
    },
  ],
});

export const getLineStyle = ({
  winner,
  length,
  height,
  left,
  heightAnimation,
  opacityAnimation,
  shakeAnimation,
}: {
  winner: PlayerMark;
  length: number;
  height: number;
  left: number;
  heightAnimation: Animated.Value;
  opacityAnimation: Animated.Value;
  shakeAnimation: Animated.Value;
}) => [
  styles.line,
  {
    backgroundColor: getPlayerColor(winner),
    width: length,
    left: left,
    height: heightAnimation,
    opacity: opacityAnimation,
  },
  getShakeStyle(shakeAnimation, length, height),
];
