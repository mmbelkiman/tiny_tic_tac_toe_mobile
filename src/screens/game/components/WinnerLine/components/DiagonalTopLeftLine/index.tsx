import React from 'react';
import {Animated} from 'react-native';
import {getLineStyle} from './utils';

interface DiagonalTopLeftLineProps {
  length: number;
  height: number;
  color: string;
  heightAnimation: Animated.Value;
  opacityAnimation: Animated.Value;
  shakeAnimation: Animated.Value;
  testID?: string | number;
}

const DiagonalTopLeftLine: React.FC<DiagonalTopLeftLineProps> = ({
  length,
  height,
  color,
  opacityAnimation,
  shakeAnimation,
  heightAnimation,
  testID = '',
}) => {
  return (
    <Animated.View
      testID={`DiagonalTopLeftLine${testID}`}
      style={getLineStyle({
        color,
        length,
        height,
        left: height,
        heightAnimation,
        opacityAnimation,
        shakeAnimation,
      })}
    />
  );
};

export default DiagonalTopLeftLine;
