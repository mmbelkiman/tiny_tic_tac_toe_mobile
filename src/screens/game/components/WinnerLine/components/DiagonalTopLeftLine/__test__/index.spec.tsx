import React from 'react';
import renderer from 'react-test-renderer';
import DiagonalTopLeftLine from '../index';
import {Animated} from 'react-native';

describe('DiagonalTopLeftLine Component', () => {
  it('should match the snapshot with a winner', () => {
    const heightAnimation = new Animated.Value(0);
    const opacityAnimation = new Animated.Value(0);
    const shakeAnimation = new Animated.Value(0);

    const tree = renderer
      .create(
        <DiagonalTopLeftLine
          length={100}
          height={20}
          color={'red'}
          heightAnimation={heightAnimation}
          opacityAnimation={opacityAnimation}
          shakeAnimation={shakeAnimation}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match the snapshot without a winner', () => {
    const heightAnimation = new Animated.Value(0);
    const opacityAnimation = new Animated.Value(0);
    const shakeAnimation = new Animated.Value(0);

    const tree = renderer
      .create(
        <DiagonalTopLeftLine
          length={200}
          height={10}
          color={'blue'}
          heightAnimation={heightAnimation}
          opacityAnimation={opacityAnimation}
          shakeAnimation={shakeAnimation}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
