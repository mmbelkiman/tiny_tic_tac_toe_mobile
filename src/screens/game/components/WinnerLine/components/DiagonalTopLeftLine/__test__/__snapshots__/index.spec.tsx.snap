// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiagonalTopLeftLine Component should match the snapshot with a winner 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "red",
      "borderRadius": 10,
      "height": 0,
      "left": 20,
      "opacity": 0,
      "position": "absolute",
      "transform": [
        {
          "translateX": -50,
        },
        {
          "translateY": -10,
        },
        {
          "rotate": "45deg",
        },
        {
          "translateX": 50,
        },
        {
          "translateY": 10,
        },
        {
          "translateY": 0,
        },
      ],
      "width": 100,
    }
  }
  testID="DiagonalTopLeftLine"
/>
`;

exports[`DiagonalTopLeftLine Component should match the snapshot without a winner 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "blue",
      "borderRadius": 10,
      "height": 0,
      "left": 10,
      "opacity": 0,
      "position": "absolute",
      "transform": [
        {
          "translateX": -100,
        },
        {
          "translateY": -5,
        },
        {
          "rotate": "45deg",
        },
        {
          "translateX": 100,
        },
        {
          "translateY": 5,
        },
        {
          "translateY": 0,
        },
      ],
      "width": 200,
    }
  }
  testID="DiagonalTopLeftLine"
/>
`;
